### DISCUSSION TITLE
Webrtc falls behind

### DISCUSSION THREAD

**Discussion by charlieopp** [bluenviron/mediamtx#1612] [Questions]
Hi, I'm using rtsp-simple-server to provide a secure webrtc interface in front of a security camera. I'm finding that as time goes by, the stream falls behind from 2 seconds when everything starts up to 15-20 seconds after running for say 8 hours. Restarting on the client doesn't help. systemctl restart rtsp-simple-server also doesn't help. So far the only way I've found to fix this is to restart the RPI on which the rtsp-simple-server is running. Any ideas on how to keep close to real-time?

Thanks,
Charlie O

**Comment by charlieopp** [comment]
For clarification, the way that I know the stream is falling behind is that the securitycam displays a timestamp. So I just look at wall time vs the displayed timestamp in the video. The spread between those increases the longer this runs. The problem is obviously that someplace in the chain, there's an expanding buffer of video and so then my question becomes "how and where could I skip to the most recent frame?"

**Comment by charlieopp** [comment]
And one more clarification. This is my ffmpeg config: ffmpeg -i rtsp://admin:xxx@10.z.x.yyy:554/h264Preview_01_sub -c:v libx264 -preset ultrafast -tune zerolatency -b:v 600k -f >

**Comment by aler9** [comment]
Hello, the server doesn't add any latency and works in FIFO mode (first input, first output). The same applies to WebRTC. Latency is probably accumulated by FFmpeg during the encoding process.

Since your camera already uses the H264 codec, re-encoding with FFmpeg is probably not necessary and you can insert the RTSP URL directly into the server configuration:

```yml
paths:
  cam:
    source: rtsp://admin:[xxx@10.z.x.yyy](mailto:xxx@10.z.x.yyy):554/h264Preview_01_sub
```

If you really need to use FFmpeg on a Raspberry Pi, my advice is to compile FFmpeg with support for hardware encoding, as explained here:
https://www.willusher.io/general/2020/11/15/hw-accel-encoding-rpi4

You can then replace libx264 with `h264_v4l2m2m`:

```
ffmpeg -i rtsp://admin:xxx@10.z.x.yyy:554/h264Preview_01_sub -c:v h264_v4l2m2m -b:v 600k -f rtsp rtsp://localhost:8554/cam
```

**Comment by charlieopp** [comment]
Thanks for this suggestion...which I've implemented and yet I still see the problem.  This is an 8GB RPI4 with nothing else running on it. Most resources idle.

As of this AM, the stream had gotten more than 15 seconds behind as viewed from my PC. I know how far behind because if I view the camera directly on another device, I can see from the timestamps in the video that it always remains essentially real-time. 

As an experiment, I disconnected the client (closed the browser tab) for two minutes, cleared the browser cache for the last hour and then connected via another tab. The stream was still behind. No change. This would suggest to me that the problem is not on the client side.

It seems to me that there's some process on the server side that dumps the cache buffer at some interval. If I remain disconnected for an hour and come back, the stream will still be behind but it won't start me going back an hour so there's some interval that the server is truncating the buffer. Is there a way that I can influence that? Or alternatively, is there a way to get the client to skip ahead in the buffer?

Thanks again,
Charlie

**Comment by charlieopp** [comment]
I believe I have now isolated the problem: It's the camera that is falling behind. This is why neither a restart of the client nor server fixes the problem. It also explains why if I let it run long-enough, the video stops. The hacky-McWork-around is that I now periodically change the source to a nonexistent source for 1 second and then set it back to the proper source. This appears to cause the buffer in the camera to flush and we get back to real-time. Doing this every 30 minutes keeps the stream always real-time.  Which leads me to a compliment to you: Thanks for allowing the server to receive a configuration update while running so that a stop - change - start cycle is not required which would add more interruption time to the video.

Thanks for the help.

Charlie

[a comment by an automated account is not shown]
